WebAppClassLoader race condition on definePackage

### GlassFish Version (and build number)

any

### JDK version

any

### OS

any

### Database

_No response_

### Problem Description

Seen in TCK CDI, failed one test of more than 60 000 executed several times. However the race condition is visible in sources quite well.
```
Caused by: java.lang.annotation.AnnotationFormatError: java.lang.IllegalArgumentException: org.testng.annotations
	at java.base/sun.reflect.annotation.AnnotationParser.parseAnnotations(AnnotationParser.java:78)
	at java.base/java.lang.Class.createAnnotationData(Class.java:4068)
	at java.base/java.lang.Class.annotationData(Class.java:4057)
	at java.base/java.lang.Class.getAnnotation(Class.java:3940)
	at java.base/java.lang.reflect.AnnotatedElement.isAnnotationPresent(AnnotatedElement.java:292)
	at java.base/java.lang.Class.isAnnotationPresent(Class.java:3950)
	at org.jboss.weld.util.Beans.isVetoed(Beans.java:520)
	at org.jboss.weld.bootstrap.AnnotatedTypeLoader.internalLoadAnnotatedType(AnnotatedTypeLoader.java:97)
	at org.jboss.weld.bootstrap.AnnotatedTypeLoader.loadAnnotatedType(AnnotatedTypeLoader.java:78)
	at org.jboss.weld.bootstrap.AnnotatedTypeLoader.loadAnnotatedType(AnnotatedTypeLoader.java:65)
	at org.jboss.weld.bootstrap.BeanDeployer.addClass(BeanDeployer.java:87)
	at org.jboss.weld.bootstrap.ConcurrentBeanDeployer$1.doWork(ConcurrentBeanDeployer.java:55)
	at org.jboss.weld.bootstrap.ConcurrentBeanDeployer$1.doWork(ConcurrentBeanDeployer.java:52)
	at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:62)
	at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:55)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	... 1 more
Caused by: java.lang.IllegalArgumentException: org.testng.annotations
	at java.base/java.lang.ClassLoader.definePackage(ClassLoader.java:2215)
	at org.glassfish.main.jdke@7.1.0-SNAPSHOT/org.glassfish.main.jdke.cl.GlassfishUrlClassLoader.definePackage(GlassfishUrlClassLoader.java:134)
	at org.glassfish.web.loader.WebappClassLoader.findClassInternal(WebappClassLoader.java:1256)
	at org.glassfish.web.loader.WebappClassLoader.findClass(WebappClassLoader.java:577)
	at org.glassfish.web.loader.WebappClassLoader.loadClass(WebappClassLoader.java:932)
	at org.glassfish.web.loader.WebappClassLoader.loadClass(WebappClassLoader.java:850)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:467)
	at java.base/sun.reflect.generics.factory.CoreReflectionFactory.makeNamedType(CoreReflectionFactory.java:114)
	at java.base/sun.reflect.generics.visitor.Reifier.visitClassTypeSignature(Reifier.java:125)
	at java.base/sun.reflect.generics.tree.ClassTypeSignature.accept(ClassTypeSignature.java:49)
	at java.base/sun.reflect.annotation.AnnotationParser.parseSig(AnnotationParser.java:441)
	at java.base/sun.reflect.annotation.AnnotationParser.parseAnnotation2(AnnotationParser.java:241)
	at java.base/sun.reflect.annotation.AnnotationParser.parseAnnotations2(AnnotationParser.java:121)
	at java.base/sun.reflect.annotation.AnnotationParser.parseAnnotations(AnnotationParser.java:73)
	... 18 more
```

We do lock the section, but just for a class name, so multiple threads can pass through finding classes of different names but of the same package.
```
        synchronized (getClassLoadingLock(name)) {
```


### Steps to reproduce

Run TCK tests many times and perhaps once you win the jackpot like me. However we do not want to see it on production. 

### Impact of Issue

_No response_

--- a/appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java
+++ b/appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2024 Contributors to the Eclipse Foundation
+ * Copyright (c) 2022, 2025 Contributors to the Eclipse Foundation.
  * Copyright (c) 1997, 2018 Oracle and/or its affiliates. All rights reserved.
  * Copyright 2004 The Apache Software Foundation
  *
@@ -47,10 +47,13 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Enumeration;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
 import java.util.function.Function;
 import java.util.jar.Attributes;
 import java.util.jar.Attributes.Name;
@@ -170,6 +173,9 @@ public final class WebappClassLoader extends GlassfishUrlClassLoader
 
     /** Associated directory context giving access to the resources in this webapp. */
     private DirContext jndiResources;
+
+    /** Maps package name to the corresponding lock object. */
+    private final Map<String, Lock> packageLocks = new ConcurrentHashMap<>();
 
     /**
      * Should this class loader delegate to the parent class loader
@@ -1147,6 +1153,8 @@ public final class WebappClassLoader extends GlassfishUrlClassLoader
             LOG.log(WARNING, "Parent close method failed.", e);
         }
 
+        packageLocks.clear();
+
         notFoundResources.clear();
         resourceEntryCache.clear();
         pathTimestamps.clear();
@@ -1244,7 +1252,7 @@ public final class WebappClassLoader extends GlassfishUrlClassLoader
             // Looking up the package
             final int pos = name.lastIndexOf('.');
             final String packageName = pos == -1 ? null : name.substring(0, pos);
-            final Package pkg;
+            Package pkg;
             if (packageName == null) {
                 pkg = null;
             } else {
@@ -1252,10 +1260,15 @@ public final class WebappClassLoader extends GlassfishUrlClassLoader
 
                 // Define the package (if null)
                 if (pkg == null) {
-                    if (entry.manifest == null) {
-                        definePackage(packageName, null, null, null, null, null, null, null);
-                    } else {
-                        definePackage(packageName, entry.manifest, entry.codeBase);
+                    Lock packageLock = getPackageDefinigLock(packageName);
+                    packageLock.lock();
+                    try {
+                        pkg = getDefinedPackage(packageName);
+                        if (pkg == null) {
+                            definePackage(packageName, entry);
+                        }
+                    } finally {
+                        packageLock.unlock();
                     }
                 }
             }
@@ -1277,6 +1290,30 @@ public final class WebappClassLoader extends GlassfishUrlClassLoader
             }
             return entry;
         }
+    }
+
+    /**
+     * Defines a Package of the given name.
+     *
+     * @param packageName the name of the to-be-defined package
+     * @param resourceEntry the resource entry
+     */
+    private void definePackage(String packageName, ResourceEntry resourceEntry) {
+        if (resourceEntry.manifest == null) {
+            definePackage(packageName, null, null, null, null, null, null, null);
+        } else {
+            definePackage(packageName, resourceEntry.manifest, resourceEntry.codeBase);
+        }
+    }
+
+    /**
+     * Returns the lock object for package defining operations.
+     *
+     * @param packageName the name of the to-be-defined package
+     * @return the lock for package defining operations
+     */
+    private Lock getPackageDefinigLock(String packageName) {
+        return packageLocks.computeIfAbsent(packageName, key -> new ReentrantLock());
     }
 
 

